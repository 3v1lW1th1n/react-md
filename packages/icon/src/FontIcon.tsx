import React, {
  forwardRef,
  FunctionComponent,
  HTMLAttributes,
  ReactNode,
  CSSProperties,
} from "react";
import cn from "classnames";
import { IWithForwardedRef } from "@react-md/utils";

const ICON_SIZE = 24;
const DENSE_ICON_SIZE = 20;

export interface IFontIconProps extends HTMLAttributes<HTMLElement> {
  /**
   * The font icon class name to use.
   */
  iconClassName?: string;

  /**
   * Boolean if the font icon should use the dense spec.
   */
  dense?: boolean;

  /**
   * Any children to render to create the font icon. This is required for material-icons.
   */
  children?: ReactNode;

  /**
   * Either a boolean that will enforce the 24x24 size of the font icon or a number of the size
   * to enforce. This is useful when using other font icon libraries that do not have a consistent
   * size.
   */
  forceSize?: boolean | number;

  /**
   * Boolean if the `forceSize` prop should also force the `font-size` instead of only `width` and
   * `height`.
   */
  forceFontSize?: boolean;
}

export interface IFontIconDefaultProps {
  dense: boolean;
  iconClassName: string;
  forceSize: boolean;
  forceFontSize: boolean;
}

export type FontIconWithDefaultProps = IFontIconProps &
  IFontIconDefaultProps &
  IWithForwardedRef;

/**
 * A utility function that will merge the different inline styles together for the `FontIcon` if
 * any of the resizing props are enabled.
 */
function createStyles(
  style: CSSProperties | undefined,
  forceSize: boolean | number,
  forceFontSize: boolean,
  dense: boolean
) {
  let nextStyle = style;
  if (typeof forceSize === "boolean" && forceSize) {
    const size = dense ? DENSE_ICON_SIZE : ICON_SIZE;
    nextStyle = {
      fontSize: forceFontSize ? size : undefined,
      height: size,
      width: size,
      ...style,
    };
  } else if (typeof forceSize === "number") {
    nextStyle = {
      fontSize: forceFontSize ? forceSize : undefined,
      height: forceSize,
      width: forceSize,
      ...style,
    };
  }

  return nextStyle;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's
 * icon. The default is to use the `material-icons` library, but others
 * can be used as well.
 *
 * If you are using another font icon library that does not always create icons with
 * a perfect 1:1 scale (such as font awesome), it is recommended to use the `forceSize`
 * and `forceFontSize` props to fix the sizing issues.
 */
const FontIcon: FunctionComponent<
  IFontIconProps & IWithForwardedRef
> = providedProps => {
  const {
    style,
    className,
    iconClassName,
    dense,
    forceSize,
    forceFontSize,
    children,
    forwardedRef,
    ...props
  } = providedProps as FontIconWithDefaultProps;

  return (
    <i
      {...props}
      ref={forwardedRef}
      style={createStyles(style, forceSize, forceFontSize, dense)}
      className={cn(
        "rmd-icon rmd-icon--font",
        {
          "rmd-icon--dense": dense,
        },
        iconClassName,
        className
      )}
    >
      {children}
    </i>
  );
};

const defaultProps: IFontIconDefaultProps = {
  dense: false,
  iconClassName: "material-icons",
  forceSize: false,
  forceFontSize: false,
};

FontIcon.defaultProps = defaultProps;

if (process.env.NODE_ENV !== "production") {
  let PropTypes = null;
  try {
    PropTypes = require("prop-types");
  } catch (e) {}

  if (PropTypes) {
    FontIcon.propTypes = {
      className: PropTypes.string,
      iconClassName: PropTypes.string,
      dense: PropTypes.bool,
      forceSize: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]),
      forceFontSize: PropTypes.bool,
      children: PropTypes.node,
    };
  }
}

export default forwardRef<HTMLElement, IFontIconProps>((props, ref) => (
  <FontIcon {...props} forwardedRef={ref} />
));
