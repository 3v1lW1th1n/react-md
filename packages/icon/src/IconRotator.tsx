import * as React from "react";
import cn from "classnames";

export interface IIconRotatorBaseProps
  extends React.HTMLAttributes<HTMLSpanElement> {
  /**
   * An optional style to apply to the surrounding span when the `forceIconWrap` prop is enabled
   * or the children is not a single react element.
   */
  style?: React.CSSProperties;

  /**
   * An optional className to apply.
   */
  className?: string;

  /**
   * Boolean if the rotation should be animated instead of static.
   */
  animate?: boolean;

  /**
   * Boolean if the icon is currently rotated.
   */
  rotated: boolean;

  /**
   * Boolean if the child icon should be "forcefully" wrapped in a `<span>` element. This should
   * be enabled if you have a custom icon that does not pass the `className` prop down.
   */
  forceIconWrap?: boolean;
}

export interface IIconRotatorProps extends IIconRotatorBaseProps {
  /**
   * The icon that should be rotated. If this is a valid React Element, the class names will be
   * cloned into that icon, otherwise the icon will be wrapped in a span with the correct class
   * names applied.
   */
  children: React.ReactElement<any> | React.ReactNode;
}

export interface IIconRotatorDefaultProps {
  animate: boolean;
  forceIconWrap: boolean;
}

export type IconRotatorWithForwardedRef = {
  forwardedRef?: React.Ref<HTMLSpanElement>;
};

export type IconRotatorWithDefaultProps = IIconRotatorProps &
  IIconRotatorDefaultProps &
  IconRotatorWithForwardedRef;

/**
 * The `IconRotator` is a simple component that is used to rotate an icon from a one degrees
 * to another.
 */
const IconRotator: React.FunctionComponent<
  IIconRotatorProps & IconRotatorWithForwardedRef
> = providedProps => {
  const {
    style,
    className: propClassName,
    animate,
    rotated,
    children,
    forceIconWrap,
    forwardedRef,
    ...props
  } = providedProps as IconRotatorWithDefaultProps;

  const className = cn(
    "rmd-icon-rotator",
    {
      "rmd-icon-rotator--animate": animate,
      "rmd-icon-rotator--rotated": rotated,
    },
    propClassName
  );

  if (!forceIconWrap && React.isValidElement(children)) {
    const child = React.Children.only(children);
    return React.cloneElement(child, {
      className: cn(className, child.props.className),
    });
  }

  return (
    <span style={style} className={className} ref={forwardedRef} {...props}>
      {children}
    </span>
  );
};

const defaultProps: IIconRotatorDefaultProps = {
  animate: true,
  forceIconWrap: false,
};

IconRotator.defaultProps = defaultProps;

export default React.forwardRef<HTMLSpanElement, IIconRotatorProps>(
  (props, ref) => <IconRotator {...props} forwardedRef={ref} />
);
