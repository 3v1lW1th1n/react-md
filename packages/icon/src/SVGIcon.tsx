import * as React from "react";
import * as PropTypes from "prop-types";
import cn from "classnames";
import memoizeOne from "memoize-one";

export interface ISVGIconProps extends React.HTMLAttributes<SVGElement> {
  /**
   * The role to apply to the SVG. When using icons, it is generally recommended to leave it as the default
   * `img` so that it is insured as a graphic.
   *
   * @docgen
   */
  role?: "img" | "presentation";

  /**
   * An optional list of ids to use to label the SVG icon with. This is helpful to add when you use the `title`
   * and `desc` props as this is used to create ids for those two props. This is super beneficial to screen readers.
   *
   * When this is defined, it is a space-delimited string of ids to provide to the title and desc (in order). If
   * this is omitted and the `use` prop is defined, it will take everything after the `#` sign and append `-title` and
   * `-desc` as a fallback. Check out the examples for more information about this.
   *
   * @see {@link #title}
   * @see {@link #desc}
   * @docgen
   */
  "aria-labelledby"?: string;

  /**
   * An optional title to give to your SVG icon. This is generally recommended for accessibility when not using
   * the `use` prop, or your spritemap does not contain `<title>` and `<desc>.
   *
   * @see {@link #aria-labelledby}
   * @docgen
   */
  title?: string;

  /**
   * An optional description to give to your SVG icon. This is generally recommended for accessibility when not using
   * the `use` prop, or your spritemap does not contain `<title>` and `<desc>.
   *
   * @see {@link #aria-labelledby}
   * @docgen
   */
  desc?: string;

  /**
   * Boolean if the SVG should gain the `focusable` attribute. This is disabled by default since IE11
   * and Edge actually default this to true and keyboard's will tab focus all SVGs.
   *
   * @docgen
   */
  focusable?: string;

  /**
   * An optional size to apply to the SVG. This can be used to set both the
   * `height` and `width` simultaneously. This will be provided as inline styles
   * since the `height` and `width` are normally controlled by CSS, and CSS has
   * higher precedence than the `height`/`width` attributes.
   *
   * @docgen
   */
  size?: number;

  /**
   * The viewBox attribute allows you to specify that a given set of graphics stretch to
   * fit a particular container element.
   *
   * The value of the viewBox attribute is a list of four numbers min-x, min-y, width and
   * height, separated by white space and/or a comma, which specify a rectangle in user
   * space which should be mapped to the bounds of the viewport established by the given
   * element, taking into account attribute preserveAspectRatio.
   *
   * Negative values for width or height are not permitted and a value of zero disables
   * rendering of the element.An optional viewbox for the SVG.
   *
   * For example, if the SVG element is 250 (width) by 200 (height) and you provide
   * `viewBox="0 0 25 20"`, the coordinates inside the SVG will go from the top left corner
   * (0, 0) to the bottom right (25, 20) and each unit will be worth `10px`.
   *
   * @docgen
   */
  viewBox?: string;

  /**
   * An optional xmlns string to provide. The `use` prop will not work without this prop
   * defined.
   *
   * @docgen
   */
  xmlns?: string;

  /**
   * This should be a link to a part of an SVG spritemap. So normally one of the following:
   * - `'#some-custom-svg'`
   * - `'/images/spritemap.svg#some-custom-svg'`
   *
   * This prop **should not** be used with the `children` prop as only one will be rendered.
   *
   * > NOTE: IE **does not support** external SVGs. Please see the demo for more details.
   *
   * @docgen
   */
  use?: string;

  /**
   * Boolean if the icon should use the dense spec.
   *
   * @docgen
   */
  dense?: boolean;

  /**
   * Any `<svg>` children to render to create your icon. This can not be used with the `use` prop.
   *
   * @docgen
   */
  children?: React.ReactNode;
}

export interface ISVGIconDefaultProps {
  role: string;
  focusable: string;
  xmlns: string;
  viewBox: string;
  dense: boolean;
}

/**
 * The `SVGIcon` component is used to render inline SVG icons or SVG icons in a sprite map
 * as an icon.
 */
export default class SVGIcon extends React.Component<ISVGIconProps, {}> {
  public static propTypes = {
    role: PropTypes.oneOf(["img", "presentation"]),
    "aria-labelledby": PropTypes.string,
    title: PropTypes.string,
    desc: PropTypes.string,
    use: PropTypes.string,
    children: PropTypes.node,
    focusable: PropTypes.string,
    size: PropTypes.number,
    viewBox: PropTypes.string,
    dense: PropTypes.bool,
    xmlns: PropTypes.string,
  };

  public static defaultProps: ISVGIconDefaultProps = {
    role: "img",
    focusable: "false",
    xmlns: "http://www.w3.org/2000/svg",
    viewBox: "0 0 24 24",
    dense: false,
  };

  private getStyle = memoizeOne((style?: React.CSSProperties, size?: number) => {
    if (style && size) {
      return { height: size, width: size, ...style };
    } else if (style) {
      return style;
    } else if (size) {
      return { height: size, width: size };
    }

    return undefined;
  });

  /**
   * A small helper function that will automatically generate specific ids within the icon
   * to add additional accessibility.
   */
  private getIds = memoizeOne((use?: string, labels?: string, title?: string, desc?: string) => {
    let titleId;
    let descId;
    let labelledBy;
    if (title || desc) {
      if (use) {
        const baseId = use.replace(/.*#/, "");
        titleId = `${baseId}-title`;
        descId = `${baseId}-desc`;

        if (title) {
          labelledBy = `${baseId}-title`;
        }

        if (desc) {
          labelledBy = `${labelledBy ? `${labelledBy} ` : ""}${descId}`;
        }
      } else if (labels) {
        [titleId, descId] = labels.split(" ");
      }
    }

    return { titleId, descId, labelledBy };
  });

  public render() {
    const {
      style,
      className,
      use,
      "aria-labelledby": ariaLabelledBy,
      size,
      title: propTitle,
      desc: propDesc,
      style: propStyle,
      children: propChildren,
      dense,
      ...props
    } = this.props;
    const { labelledBy, titleId, descId } = this.getIds(use, ariaLabelledBy, propTitle, propDesc);

    let title = null;
    let desc = null;
    let children = propChildren;
    if (!children && use) {
      children = <use xlinkHref={use} />;
    }

    if (propTitle) {
      title = <title id={titleId}>{propTitle}</title>;
    }

    if (desc) {
      desc = <desc id={descId}>{desc}</desc>;
    }

    return (
      <svg
        {...props}
        style={this.getStyle(style, size)}
        aria-labelledby={ariaLabelledBy || labelledBy}
        className={cn(
          "rmd-icon rmd-icon--svg",
          {
            "rmd-icon--svg-dense": dense,
          },
          className
        )}
      >
        {title}
        {desc}
        {children}
      </svg>
    );
  }
}
