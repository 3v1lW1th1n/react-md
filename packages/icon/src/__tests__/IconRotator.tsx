import * as React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";

import IconRotator from "../IconRotator";
import FontIcon from "../FontIcon";
import SVGIcon from "../SVGIcon";

const Icon = ({ className }: any) => (
  <i className={`${className} custom-icon`} />
);

describe("IconRotator", () => {
  it("should default to cloning the icon rotator classNames into the child element", () => {
    const icon = mount(
      <IconRotator rotated={false}>
        <Icon />
      </IconRotator>
    );

    expect(icon.find("i").hasClass("custom-icon")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon-rotator")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon-rotator--animate")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon-rotator--rotated")).toBe(false);
    expect(icon.render()).toMatchSnapshot();

    icon.setProps({ rotated: true });
    expect(icon.find("i").hasClass("custom-icon")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon-rotator")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon-rotator--animate")).toBe(true);
    expect(icon.find("i").hasClass("rmd-icon-rotator--rotated")).toBe(true);
    expect(icon.render()).toMatchSnapshot();
  });

  it("should be able to work out of the box with SVGIcon and FontIcon", () => {
    const fontRotator = mount(
      <IconRotator rotated={false}>
        <FontIcon>home</FontIcon>
      </IconRotator>
    );
    const svgRotator = mount(
      <IconRotator rotated={false}>
        <SVGIcon>
          <path d="0i3odksf" />
        </SVGIcon>
      </IconRotator>
    );

    expect(fontRotator.render()).toMatchSnapshot();
    expect(svgRotator.render()).toMatchSnapshot();

    fontRotator.setProps({ rotated: true });
    svgRotator.setProps({ rotated: true });
    expect(fontRotator.render()).toMatchSnapshot();
    expect(svgRotator.render()).toMatchSnapshot();
  });

  it("should wrap the children with a span instead of cloning in the className when the forceIconWrap prop is enabled", () => {
    expect(
      create(
        <IconRotator rotated={false} forceIconWrap>
          <Icon />
        </IconRotator>
      ).toJSON()
    ).toMatchSnapshot();
    expect(
      create(
        <IconRotator rotated={false} forceIconWrap>
          <FontIcon>home</FontIcon>
        </IconRotator>
      ).toJSON()
    ).toMatchSnapshot();
    expect(
      create(
        <IconRotator rotated={false} forceIconWrap>
          <SVGIcon>
            <path d="0i3odksf" />
          </SVGIcon>
        </IconRotator>
      ).toJSON()
    ).toMatchSnapshot();
  });
});
