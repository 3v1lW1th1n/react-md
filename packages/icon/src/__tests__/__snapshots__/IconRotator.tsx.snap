// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconRotator should be able to work out of the box with SVGIcon and FontIcon 1`] = `
<i
  aria-hidden="true"
  class="rmd-icon rmd-icon--font material-icons rmd-icon-rotator rmd-icon-rotator--animate"
>
  home
</i>
`;

exports[`IconRotator should be able to work out of the box with SVGIcon and FontIcon 2`] = `
<svg
  aria-hidden="true"
  class="rmd-icon rmd-icon--svg rmd-icon-rotator rmd-icon-rotator--animate"
  focusable="false"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="0i3odksf"
  />
</svg>
`;

exports[`IconRotator should be able to work out of the box with SVGIcon and FontIcon 3`] = `
<i
  aria-hidden="true"
  class="rmd-icon rmd-icon--font material-icons rmd-icon-rotator rmd-icon-rotator--animate rmd-icon-rotator--rotated"
>
  home
</i>
`;

exports[`IconRotator should be able to work out of the box with SVGIcon and FontIcon 4`] = `
<svg
  aria-hidden="true"
  class="rmd-icon rmd-icon--svg rmd-icon-rotator rmd-icon-rotator--animate rmd-icon-rotator--rotated"
  focusable="false"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="0i3odksf"
  />
</svg>
`;

exports[`IconRotator should default to cloning the icon rotator classNames into the child element 1`] = `
<i
  class="rmd-icon-rotator rmd-icon-rotator--animate custom-icon"
/>
`;

exports[`IconRotator should default to cloning the icon rotator classNames into the child element 2`] = `
<i
  class="rmd-icon-rotator rmd-icon-rotator--animate rmd-icon-rotator--rotated custom-icon"
/>
`;

exports[`IconRotator should wrap the children with a span instead of cloning in the className when the forceIconWrap prop is enabled 1`] = `
<span
  className="rmd-icon-rotator rmd-icon-rotator--animate"
>
  <i
    className="undefined custom-icon"
  />
</span>
`;

exports[`IconRotator should wrap the children with a span instead of cloning in the className when the forceIconWrap prop is enabled 2`] = `
<span
  className="rmd-icon-rotator rmd-icon-rotator--animate"
>
  <i
    aria-hidden="true"
    className="rmd-icon rmd-icon--font material-icons"
  >
    home
  </i>
</span>
`;

exports[`IconRotator should wrap the children with a span instead of cloning in the className when the forceIconWrap prop is enabled 3`] = `
<span
  className="rmd-icon-rotator rmd-icon-rotator--animate"
>
  <svg
    aria-hidden="true"
    className="rmd-icon rmd-icon--svg"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="0i3odksf"
    />
  </svg>
</span>
`;
